src/p2pool/P2PoolManager.cpp

```diff
@@ -39,6 +41,96 @@
 #include <QApplication>
 #include <QProcess>
 #include <QMap>
+#include <QCryptographicHash>
+
+void P2PoolManager::download() {
+    m_scheduler.run([this] {
+        QUrl url;
+        QString fileName;
+        #ifdef Q_OS_WIN
+            url = "https://github.com/SChernykh/p2pool/releases/download/v1.6/p2pool-v1.6-windows-x64.zip";
+            fileName = "p2pool-v1.6-windows-x64.zip";
+        #elif defined(Q_OS_LINUX)
+            url = "https://github.com/SChernykh/p2pool/releases/download/v1.6/p2pool-v1.6-linux-x64.tar.gz";
+            fileName = "p2pool-v1.6-linux-x64.tar.gz";
+        #elif defined(Q_OS_MACOS)
+            url = "https://github.com/SChernykh/p2pool/releases/download/v1.6/p2pool-v1.6-macos-x64.tar.gz";
+            fileName = "p2pool-v1.6-macos-x64.tar.gz";
+        #endif
+        QFile file(fileName);
+        epee::net_utils::http::http_simple_client http_client;
+        const epee::net_utils::http::http_response_info* response = NULL;
+        std::string userAgent = randomUserAgent().toStdString();
+        std::chrono::milliseconds timeout = std::chrono::seconds(10);
+        http_client.set_server(url.host().toStdString(), "443", {});
+        bool success = http_client.invoke_get(url.path().toStdString(), timeout, {}, std::addressof(response), {{"User-Agent", userAgent}});
+        if (response->m_response_code == 302) {
+            epee::net_utils::http::fields_list fields = response->m_header_info.m_etc_fields;
+            for (std::pair<std::string, std::string> i : fields) {
+                if (i.first == "Location") {
+                    url = QString::fromStdString(i.second);
+                    http_client.set_server(url.host().toStdString(), "443", {});
+                    std::string query = url.query(QUrl::FullyEncoded).toStdString();
+                    std::string path = url.path().toStdString() + "?" + query;
+                    http_client.wipe_response();
+                    success = http_client.invoke_get(path, timeout, {}, std::addressof(response), {{"User-Agent", userAgent}});
+                }
+            }
+        }
+        if (!success) {
+            emit p2poolDownloadFailure();
+        }
+        else {
+            std::string stringData = response->m_body;
+            QByteArray data(stringData.c_str(), stringData.length());
+            QByteArray hashData = QCryptographicHash::hash(data, QCryptographicHash::Sha256);
+            QString hash = hashData.toHex();
+            if (fileName == "p2pool-v1.6-windows-x64.zip") {
+                if (hash != "d38493ab7549bd146672afb576279d51ea8fc7f5e75fd28ecabaa4c1f97644a6") {
+                    emit p2poolDownloadFailure();
+                }
+                else if (hash == "d38493ab7549bd146672afb576279d51ea8fc7f5e75fd28ecabaa4c1f97644a6") { 
```

Comment:
This check is redundant, the same hash is checked a few lines up (same for other files).